mithril-common/src/chain_observer/pallas_observer.rs

```diff
@@ -0,0 +1,238 @@
+
+        statequery.acquire(None).await.unwrap();
+
+        let era = queries_v16::get_current_era(statequery)
+            .await
+            .map_err(|err| ChainObserverError::General(err.into()))?;
+
+        let epoch = queries_v16::get_block_epoch_number(statequery, era)
+            .await
+            .map_err(|err| ChainObserverError::General(err.into()))?;
+
+        statequery.send_release().await.unwrap();
+        statequery.send_done().await.unwrap();
+
+        client.chainsync().send_done().await.unwrap();
+        drop(client.plexer_handle);
+
+        Ok(Some(Epoch(epoch as u64)))
+    }
+
+    async fn get_current_datums(
+        &self,
+        address: &ChainAddress,
+    ) -> Result<Vec<TxDatum>, ChainObserverError> {
+        let fallback = self.get_fallback().map_err(ChainObserverError::General)?;
+        fallback.get_current_datums(address).await
+    }
+
+    async fn get_current_stake_distribution(
+        &self,
+    ) -> Result<Option<StakeDistribution>, ChainObserverError> {
+        let fallback = self.get_fallback().map_err(ChainObserverError::General)?;
+        fallback.get_current_stake_distribution().await
+    }
+
+    async fn get_current_kes_period(
+        &self,
+        opcert: &OpCert,
+    ) -> Result<Option<KESPeriod>, ChainObserverError> {
+        let fallback = self.get_fallback().map_err(ChainObserverError::General)?;
+        fallback.get_current_kes_period(opcert).await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::fs;
+
+    use pallas_codec::utils::AnyCbor;
+    use pallas_network::miniprotocols::{
+        chainsync::{self},
+        localstate::{self, queries_v16::EpochNo, ClientQueryRequest},
+    };
+    use tokio::net::UnixListener;
+
+    use super::*;
+    use crate::CardanoNetwork;
+
+    #[tokio::test]
+    #[ignore]
+    async fn get_current_epoch_with_fallback() {
+        let server = tokio::spawn({
+            async move {
+                let socket_path = Path::new("node.socket");
+                if socket_path.exists() {
+                    fs::remove_file(socket_path).unwrap();
+                }
+                let unix_listener = UnixListener::bind(socket_path).unwrap();
+
+                let mut server = pallas_network::facades::NodeServer::accept(&unix_listener, 10)
+                    .await
+                    .unwrap();
+
+                let maybe_acquire = server.statequery().recv_while_idle().await.unwrap();
+                assert!(maybe_acquire.is_some());
+                assert_eq!(*server.statequery().state(), localstate::State::Acquiring);
+
+                server.statequery().send_acquired().await.unwrap();
+                assert_eq!(*server.statequery().state(), localstate::State::Acquired);
+
+                //////////////////////////////////////////////////////////////
+                //                        Get Current Era                   //
+                //////////////////////////////////////////////////////////////
```

Comment:
Please don't use header-style comments please. If you need to isolate a block of code, abstract it away into it's own function and add a docstring to it. If you're worried about the penalty of having an extra function, use `#[inline]`

Reply:
done @scarmuega